src/java/org/apache/cassandra/service/snapshot/SnapshotWatcher.java

```diff
@@ -0,0 +1,186 @@
+ * in SnapshotManager anymore hence nodetool listsnapshots nor system_views.snapshots will not display it either.
+ */
+public class SnapshotWatcher implements AutoCloseable
+{
+    private static final Logger logger = LoggerFactory.getLogger(SnapshotWatcher.class);
+
+    private final SnapshotManager snapshotManager;
+    private WatchService watchService;
+    private ExecutorService executor;
+    private final Map<WatchKey, Path> watchKeyPathMap = new ConcurrentHashMap<>();
+
+    private Future<?> watcherFuture;
+
+    public SnapshotWatcher(SnapshotManager snapshotManager)
+    {
+        this.snapshotManager = snapshotManager;
+    }
+
+    public void watch(TableSnapshot snapshot)
+    {
+        Collection<File> directories = snapshot.getDirectories();
+        if (directories == null)
+            return;
+
+        for (File snapshotDir : directories)
+        {
+            Path snapshotPath = snapshotDir.toPath();
+            try
+            {
+                WatchKey watchKey = snapshotPath.register(watchService, ENTRY_DELETE);
+                watchKeyPathMap.put(watchKey, snapshotPath);
+            }
+            catch (IOException ex)
+            {
+                logger.warn("Unable to register watch service for path " + snapshotPath, ex);
+            }
+        }
+    }
+
+    public void unwatch(TableSnapshot snapshot)
+    {
+        for (File snapshotDir : snapshot.getDirectories())
+        {
+            Path snapshotPath = snapshotDir.toPath();
+            for (Map.Entry<WatchKey, Path> entry : watchKeyPathMap.entrySet())
+            {
+                if (entry.getValue().equals(snapshotPath))
+                {
+                    WatchKey watchKey = entry.getKey();
+                    watchKeyPathMap.remove(watchKey);
+                }
+            }
+        }
+    }
+
+    public void start()
+    {
+        if (watchService != null || executor != null)
+            return;
+
+        try
+        {
+            watchService = FileSystems.getDefault().newWatchService();
+        }
+        catch (IOException ex)
+        {
+            throw new RuntimeException("Unable to get watch service", ex);
+        }
+
+        executor = getExecutorService();
+
+        watcherFuture = executor.submit(() -> {
+            while (!Thread.currentThread().isInterrupted())
+            {
+                WatchKey key;
+
+                try
+                {
+                    key = watchService.take();
+
+                    for (WatchEvent<?> event : key.pollEvents())
+                    {
+                        if (event.kind() == ENTRY_DELETE)
+                        {
+                            Path dir = watchKeyPathMap.get(key);
+
+                            if (dir == null)
+                                continue;
+
+                            snapshotManager.findSnapshot(dir).ifPresent(snapshotManager::clearSnapshot);
```

Comment:
This should not block and clearing of a snapshot should be done asynchronously to process events as fast as possible as doing it in a blocking manner might cause overflowing of deletion events under extreme conditions.

We should just populate a queue of snapshots to be cleared and then a thread in SnapshotManager should just go over it.